File: backend/app/schemas/p2.py
"""P2 Schema。"""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class BriefingResponse(BaseModel):
    briefing: str


class ReviewResponse(BaseModel):
    review: str


class ConfirmExtractions(BaseModel):
    ids: list[uuid.UUID] = Field(default_factory=list)


class PracticeSessionCreate(BaseModel):
    scenario_id: uuid.UUID | None = None
    title: str | None = None
    channel: Literal["online", "offline"] = "offline"
    tags: list[str] = Field(default_factory=list)
    custom_prompt: str | None = None
    participants: list[dict] = Field(default_factory=list)


class PracticeScenarioCreate(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    channel: Literal["online", "offline"] = "offline"
    tags: list[str] = Field(default_factory=list)
    custom_prompt: str | None = None
    participants: list[dict] = Field(default_factory=list)


class BackgroundGenerate(BaseModel):
    channel: Literal["online", "offline"] = "offline"
    tags: list[str] = Field(default_factory=list)
    custom_prompt: str | None = None
    person_ids: list[uuid.UUID] = Field(default_factory=list)
    roles: list[dict] = Field(default_factory=list)


class PracticeSend(BaseModel):
    content: str = Field(min_length=1)


class MemoryCreate(BaseModel):
    kind: Literal["preference", "goal", "habit", "other"] = "preference"
    content: str = Field(min_length=1)
    person_id: uuid.UUID | None = None


class MemoryUpdate(BaseModel):
    status: Literal["accepted", "ignored"] | None = None
    content: str | None = None


class UserProfileRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    expression_preferences: dict | None
    social_goals: dict | None


class UserProfileUpdate(BaseModel):
    expression_preferences: dict | None = None
    social_goals: dict | None = None


class DashboardData(BaseModel):
    persons: int = 0
    interactions: int = 0
    topics: int = 0
    documents: int = 0
    due_followups: list[dict] = Field(default_factory=list)
    due_reviews: list[dict] = Field(default_factory=list)
    recent_interactions: list[dict] = Field(default_factory=list)

File: frontend/src/api/p2.ts
import { request } from "@/api/client";

export function generateBriefing(personId: string): Promise<string> {
  return request<{ briefing: string }>(`/api/persons/${personId}/briefing`, {
    method: "POST",
  }).then((data) => data.briefing);
}

export function extractInteraction(interactionId: string): Promise<ExtractionItem[]> {
  return request<ExtractionItem[]>(`/api/interactions/${interactionId}/extract`, {
    method: "POST",
  });
}

export interface ExtractionItem {
  id: string;
  kind: string;
  fact_type: string;
  content: string;
  status: string;
}

export function getPendingExtractions(interactionId: string): Promise<ExtractionItem[]> {
  return request<ExtractionItem[]>(`/api/interactions/${interactionId}/extractions`);
}

export function confirmExtractions(
  interactionId: string,
  ids: string[],
): Promise<{ confirmed: number }> {
  return request(`/api/interactions/${interactionId}/confirm-extractions`, {
    method: "POST",
    body: JSON.stringify({ ids }),
  });
}

export function reviewInteraction(interactionId: string): Promise<string> {
  return request<{ review: string }>(`/api/interactions/${interactionId}/review`, {
    method: "POST",
  }).then((data) => data.review);
}

export interface Scenario {
  id: string;
  scenario_type: string;
  title: string;
  description: string | null;
  channel: string;
  tags: string[];
  custom_prompt: string | null;
  participants: { name: string; role?: string; person_id?: string }[];
}

export function listScenarios(): Promise<Scenario[]> {
  return request<Scenario[]>("/api/practice/scenarios");
}

export interface TagLibrary {
  场合: string[];
  谈话背景: string[];
  对象类型: string[];
}

export function listTagLibrary(): Promise<TagLibrary> {
  return request<TagLibrary>("/api/practice/tag-library");
}

export function generateBackground(input: {
  channel: string;
  tags?: string[];
  custom_prompt?: string | null;
  person_ids?: string[];
  roles?: { name: string; role?: string }[];
}): Promise<string> {
  return request<{ background: string }>("/api/practice/generate-background", {
    method: "POST",
    body: JSON.stringify(input),
  }).then((data) => data.background);
}

export function createCustomScenario(input: {
  title: string;
  channel: string;
  tags?: string[];
  custom_prompt?: string | null;
  participants?: { name: string; role?: string }[];
}): Promise<{ id: string; title: string; channel: string; tags: string[] }> {
  return request("/api/practice/scenarios", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function createPracticeSession(
  scenarioId: string | null,
  options?: {
    channel?: string;
    tags?: string[];
    custom_prompt?: string | null;
    participants?: { name: string; role?: string; person_id?: string }[];
  },
): Promise<{
  id: string;
  title: string;
  channel: string;
  tags: string[];
  custom_prompt: string | null;
  participants: { name: string; role?: string }[];
}> {
  return request("/api/practice/sessions", {
    method: "POST",
    body: JSON.stringify({ scenario_id: scenarioId, ...options }),
  });
}

export interface PracticeSessionInfo {
  id: string;
  title: string;
  status: string;
  channel: string;
  tags: string[];
  custom_prompt: string | null;
  participants: { name: string; role?: string; person_id?: string }[];
  created_at: string;
}

export function listPracticeSessions(): Promise<PracticeSessionInfo[]> {
  return request<PracticeSessionInfo[]>("/api/practice/sessions");
}

export function deletePracticeSession(sessionId: string): Promise<void> {
  return request(`/api/practice/sessions/${sessionId}`, { method: "DELETE" });
}

export interface PracticeMessageItem {
  id: string;
  role: string;
  content: string;
}

export function listPracticeMessages(sessionId: string): Promise<PracticeMessageItem[]> {
  return request<PracticeMessageItem[]>(`/api/practice/sessions/${sessionId}/messages`);
}

export function practiceSendStream(sessionId: string, content: string): Promise<Response> {
  return fetch(`/api/practice/sessions/${sessionId}/messages`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
}

export function evaluateSession(
  sessionId: string,
): Promise<{ scores: Record<string, number>; summary: string }> {
  return request(`/api/practice/sessions/${sessionId}/evaluate`, { method: "POST" });
}

export function suggestReplies(sessionId: string): Promise<string[]> {
  return request<{ suggestions: string[] }>(
    `/api/practice/sessions/${sessionId}/suggest-replies`,
    { method: "POST" },
  ).then((data) => data.suggestions);
}

export interface ReviewItem {
  id: string;
  topic_id: string;
  topic_name: string;
  due_at: string;
  interval_days: number;
}

export function listDueReviews(): Promise<ReviewItem[]> {
  return request<ReviewItem[]>("/api/reviews/due");
}

export function answerReview(taskId: string, rating: string): Promise<{ interval_days: number }> {
  return request(`/api/reviews/${taskId}/answer?rating=${encodeURIComponent(rating)}`, {
    method: "POST",
  });
}

export interface DashboardData {
  persons: number;
  interactions: number;
  topics: number;
  documents: number;
  due_followups: { id: string; title: string; person_name: string }[];
  due_reviews: ReviewItem[];
  recent_interactions: { id: string; title: string; occurred_at: string }[];
}

export function getDashboard(): Promise<DashboardData> {
  return request<DashboardData>("/api/dashboard");
}

export function weeklyReport(): Promise<string> {
  return request<{ report: string }>("/api/reports/weekly", { method: "POST" }).then(
    (data) => data.report,
  );
}

export interface RelationshipItem {
  id: string;
  other_person_id: string;
  other_person_name: string;
  relation_type: string;
  note: string | null;
}

export function listRelationships(personId: string): Promise<RelationshipItem[]> {
  return request<RelationshipItem[]>(`/api/persons/${personId}/relationships`);
}

export function addRelationship(
  personId: string,
  input: { other_person_id: string; relation_type: string; note?: string | null },
): Promise<RelationshipItem> {
  return request<RelationshipItem>(`/api/persons/${personId}/relationships`, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function deleteRelationship(relationshipId: string): Promise<void> {
  return request(`/api/relationships/${relationshipId}`, { method: "DELETE" });
}
